core/workflow-core/src/main/scala/edu/uci/ics/amber/core/storage/result/MemoryDocument.scala

```diff
@@ -0,0 +1,107 @@
+package edu.uci.ics.amber.core.storage.result
+
+import edu.uci.ics.amber.core.storage.model.{BufferedItemWriter, VirtualDocument}
+
+import java.net.URI
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+/**
+  * Global storage mapping for all MemoryDocuments.
+  * This ensures that 2 MemoryDocuments opened on the same key will operate on the same ArrayBuffer.
+  */
+object MemoryDocument {
+  val storageMapping: mutable.Map[String, ArrayBuffer[AnyRef]] =
+    mutable.HashMap[String, ArrayBuffer[AnyRef]]()
+}
+
+/**
+  * MemoryDocument provides an in-memory implementation of VirtualDocument and BufferedItemWriter.
+  * It stores items in a mutable ArrayBuffer and supports various operations like read, write, and remove.
+  * @tparam T the type of data items stored in the document.
+  */
+class MemoryDocument[T >: Null <: AnyRef](key: String)
+  extends VirtualDocument[T]
+    with BufferedItemWriter[T] {
+
+  /**
+    * Internal storage for the document's items, retrieved from the shared mapping.
+    */
+  private def results: ArrayBuffer[T] = {
+    if (!MemoryDocument.storageMapping.contains(key)) {
+      MemoryDocument.storageMapping(key) = new ArrayBuffer[AnyRef]()
+    }
+    MemoryDocument.storageMapping(key).asInstanceOf[ArrayBuffer[T]]
+  }
+
+  override def getURI: URI =
+    throw new UnsupportedOperationException("getURI is not supported for MemoryDocument")
+
+  override def clear(): Unit =
+    synchronized {
+      results.clear()
+    }
+
+  override def get(): Iterator[T] =
+    synchronized {
+      results.to(Iterator)
+    }
+
+  override def getItem(i: Int): T =
+    synchronized {
+      results.apply(i)
+    }
+
+  override def getRange(from: Int, until: Int): Iterator[T] =
+    synchronized {
+      results.slice(from, until).to(Iterator)
+    }
+
+  override def getAfter(offset: Int): Iterator[T] =
+    synchronized {
+      results.slice(offset, results.size).to(Iterator)
+    }
+
+  override def getCount: Long = results.length
+
+  override def append(item: T): Unit =
+    synchronized {
+      results += item
+    }
+
+  override def writer(): BufferedItemWriter[T] = this
```

Comment:
all concurrent threads will get access to the same `BufferedItemWriter` instance. is this thread safe to write in multiple threads?

Reply:
Since each read/write method of MemoryDocument is protected using `synchronized` keyword, it is thread safe to write in multiple threads